RavenDB-15062 - Filtered Replication [ Review Only ]

ayende opened the pull request.

ayende committed:
RavenDB-15062 - Refactoring to reduce duplication in stream destination

ayende committed:
RavenDB-15062 Allow to setup filters on pull replication that would limit what is sent to the other side

ayende committed:
RavenDB-15062 - Added tests to verify functionality
RavenDB-15076 - Replication, Counters and Revisions do not play well together
- Made sure that ForceRevision is available in the public API as well

ayende committed:
RavenDB-15062 Adding support for push filtered replication

ayende commented on src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs:
```diff
@@ -200,7 +200,7 @@ private void Replication()
                         {
                             SendPreliminaryData();
-                            if (Destination is PullReplicationAsSink)
+                            if (Destination is PullReplicationAsSink sink && sink.Mode == ReplicationMode.Pull)
```
Note, here we don't use flags, only the real value.

ayende committed:
RavenDB-15062 Reverting unnecessary changes

ayende reviewed (commented):
This adds the capability to add filters to replication hub as well as support pull & push on the hub.
The idea is that the hub contains a list of certificates as well as the paths that each of them is allowed to write / read.
Both incoming & outgoing replication will filter data (documents, counters, time series, attachments and revisions) so only documents that match the
allowed paths will be replicated.
Note that within a document, all the data about the document is either replicated or not. There is no way to filter data inside a document.

This is a *security* feature that assume that the other side is *not* trusted. We should think about ways that people can try to mess with us here.

When using pull filtered replication, the hub will filter out all the paths that the user is not allowed to see at the source.
When using push filtered replication, the hub will let the pushing node what allowed paths it will accept, and the pusher will filter them before
sending them up. The hub will also validate that the incoming data match the allowed filters. If the pusher send data that it shouldn't, we abort
the connection and force a retry.

Filtering is only enabled for certificates registered directly on the hub. If anything has a filter, all certificates must have a filter. Note that
accessing the pull replication using already trusted certificate that isn't registered on the hub means that we don't do *any* filtering and give
unfettered access.

We now have the concept of a replication hub that can be set to: Pull, Push or Pull & Push. It can accept either a pull replication. In which case
it will work normaly, but filter the results to send or accept a push replication, in which case it will send to the pusher the allowed paths and
then validate them locally.

On the other side, we have a sink. These can be configured as either Push or Pull (never both) and will function as the other side of the hub, 
accepting data from the remote hub or pushing filtered data here.

In this context, the hub is *trusted* and the end node is *not* trusted. 

Remaining work:
* How should we have such replication when using output collections from map/reduce. 
* Optimizing pull replication for larger number of certificates.
* We use the term Pull Replication Hub throughout our code. We have now changed it to also allow pushing. How to deal with this? Same with the sink
  terminology.

ayende commented:
One of the things we need to see if we can do is to cheat. 
Can we provide one value at authorization phase and then use another during the actual replication setup?

ayende committed:
RavenDB-15071 Refactor PullReplication.Certificates - Right now we simply ensure that we don't have to materialize the whole object when we access any property in it.

karmeli87 commented on src/Raven.Client/Documents/Operations/Replication/ExternalReplication.cs:
```diff
@@ -13,12 +13,14 @@ public ExternalReplication(string database, string connectionStringName) : base(
             
         }
 
+        public string HubDefinitionName;
```
Why is it on the `ExternalReplication`?

ayende replied:
Removed

karmeli87 commented on src/Raven.Client/Documents/Operations/Replication/PullReplicationDefinition.cs:
```diff
@@ -1,23 +1,125 @@
+
+        public class FilteringOptions : IDynamicJson
+        {
+            public string[] AllowedPaths;
```
We can actually have a different filters for pull and push.
Not hard to add and I think it might be useful.
e.g. I want to pull only configuration document, but push all the data that I've collected.

ayende replied:
Good idea.

karmeli87 commented on src/Raven.Client/Documents/Operations/Replication/ReplicationMode.cs:
```diff
@@ -0,0 +1,13 @@
+        None = 0,
+        Pull = 1,
+        Push = 2,
+        Invalid = 4
```
Why do we need that?

ayende replied:
Removed

karmeli87 commented on src/Raven.Server/Documents/Replication/AllowedPathsValidator.cs:
```diff
@@ -0,0 +1,107 @@
+        public AllowedPathsValidator(string[] allowedPaths)
+        {
+            {
+                _allowedPathsContext = JsonOperationContext.ShortTermSingleUse();
```
Not familiar with the `ShortTermSingleUse`, just wanted to notice that this isn't a "short-term" usage.

ayende replied:
That just means that it isn't using a context. This will live as long as the validator lives, which is fine.

karmeli87 commented on src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs:
```diff
@@ -512,6 +512,14 @@ private void InitialHandshake()
+                        {
+                            // this is used when the other side lets us know what 
+                            // paths it is going to accept from us
+                            AllowedPaths = response.Reply.AllowedPaths;
```
Need to validate that changing the definition on the Hub will be reflected on the already opened connections

ayende replied:
Right now, IIRC, that wouldn't be the case for things like removing certificates.
Existing connections are maintained. 

But I agree: https://issues.hibernatingrhinos.com/issue/RavenDB-15090

karmeli87 commented on src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs:
```diff
@@ -535,6 +538,13 @@ private bool AddReplicationItemToBatch(ReplicationBatchItem item, OutgoingReplic
+            if (_allowedPathsValidator?.ShouldAllow(item) == false)
+            {
+                stats.RecordArtificialDocumentSkip();
+                skippedReplicationItemsInfo.Update(item);
```
Need to write a correct log message, don't want this to mislead us.

ayende replied:
done

karmeli87 commented on src/Raven.Server/Documents/Replication/ReplicationLoader.cs:
```diff
@@ -180,50 +180,101 @@ public ReplicationLoader(DocumentDatabase database, ServerStore server)
-                    if (initialRequest.PullReplicationDefinitionName != null)
+                case TcpConnectionHeaderMessage.AuthorizationInfo.AuthorizeMethod.PullReplication:
+                case TcpConnectionHeaderMessage.AuthorizationInfo.AuthorizeMethod.PushReplication:
+                    if (supportedVersions.Replication.PullReplication == false)
```
Shouldn't this `if` come before the `switch` ?

karmeli87 replied:
It is a new feature so we also need to add to the supported versions an option for push/filtered replication.

ayende replied:
No, it shouldn't. We need to do this check if the other side requested pull replication in this manner. 
There should be no way that you'll get here if you aren't supporting pull replication

karmeli87 commented on src/Raven.Server/ServerWide/RawDatabaseRecord.cs:
```diff
@@ -248,29 +250,64 @@ public List<ExternalReplication> ExternalReplications
-        private List<PullReplicationDefinition> _hubPullReplications;
-
-        public List<PullReplicationDefinition> HubPullReplications
+        public List<RawPullReplicationDefinition> RawHubPullReplications
```
Does this change has any backward comp. issues that we care about?

karmeli87 commented on src/Raven.Client/Documents/Operations/Replication/PullReplicationDefinition.cs:
```diff
@@ -1,23 +1,125 @@
+        }
+    }
+
+    public class RawPullReplicationDefinition
```
Why we have this and `PullReplicationDefinition`  ?

ayende replied:
Mid way through refactoring things to avoid putting certificates inside the database record.

ayende committed:
RavenDB-15062
- PR Comments
- Read/Write vs. Pull/Push term - to make it easier to use pull replication for writing
- Allow to define separate read / write paths
- Better error handling
- Removing commented code

ayende committed:
RavenDB-15062 - Starting to move the hub access outside of the database record.

ayende committed:
RavenDB-15062 - Moving filtered replication to store its state in the server store.

ayende committed:
RavenDB-15062
- Properly setup push / pull replication via register access commands
- Allow easier debugging of tests when running in secured mode by trusting local requests without needing to install (temporary) certificate
- Properly handle security checks for replication hub connections

ayende committed:
RavenDB-15090 Pull replication - drop connections when the configuration changes
